frontend/src/liboperator/operator/operator.go

```diff
@@ -533,60 +534,49 @@ func adbProxy(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
 	defer tcpConn.Close()
 
 	// Redirect WebSocket to ADB tcp socket
-	redirectWsToTcp(wsConn, tcpConn)
-}
-
-func redirectWsToTcp(wsConn *websocket.Conn, tcpConn net.Conn) {
-	tcpChan := make(chan []byte)
-	wsChan := make(chan []byte)
-
-	go func(tcpConn net.Conn, tcpChan chan []byte) {
-		buf := make([]byte, 4096)
-		for {
-			nRead, err := tcpConn.Read(buf)
-
-			if err != nil {
-				close(tcpChan)
-				return
-			}
-
-			tcpData := make([]byte, nRead)
-			copy(tcpData, buf)
-
-			tcpChan <- tcpData
-		}
-	}(tcpConn, tcpChan)
+	wsWrapper := &wsIoWrapper{
+		wsConn: wsConn,
+		pos: 0,
+		buf: nil,
+	}
 
-	go func(wsConn *websocket.Conn, wsChan chan []byte) {
-		for {
-			_, buf, err := wsConn.ReadMessage()
+	go io.Copy(wsWrapper, tcpConn)
+	io.Copy(tcpConn, wsWrapper)
+}
 
-			if err != nil {
-				close(wsChan)
-				return
-			}
+// Wrapper for implementing io.ReadWriteCloser of websocket.Conn
+type wsIoWrapper struct {
+	wsConn *websocket.Conn
+	pos int
+	buf []byte
+}
 
-			wsData := make([]byte, len(buf))
-			copy(wsData, buf)
+var _ io.ReadWriteCloser = (*wsIoWrapper)(nil)
 
-			wsChan <- wsData
+func (w *wsIoWrapper) Read(buf []byte) (int, error) {
+	if w.buf == nil || w.pos == len(w.buf) {
+		_, buf, err := w.wsConn.ReadMessage()
+		if err != nil {
+			return 0, err
 		}
-	}(wsConn, wsChan)
+		w.buf = buf
+		w.pos = 0
+	}
+	nRead := copy(buf[:], w.buf[w.pos:])
```

Comment:
we can use `buf` directly here as it's already an slice.

Reply:
I'll upload separated PR for this.